README.txt
pyramid_beaker
==============

Provides a session factory for the `Pyramid <http://docs.pylonsproject.org>`_
web framework backed by the `Beaker <http://beaker.groovie.org/>`_ sessioning
system.

See `the Pylons Project documentation <http://docs.pylonsproject.org>`_ for
more information.


CHANGES.txt
unreleased
----------

 - Fixed a bug causing session saving even when it is not needed. See
   https://github.com/Pylons/pyramid_beaker/pull/28

0.8 (2013-06-28)
----------------

- On Python 3, the CSRF token returned by new_csrf_token was a bytestring
  instead of correctly being a string.  See
  https://github.com/Pylons/pyramid_beaker/pull/12

- A new argument to PyramidBeakerSessionFactoryConfig named
  ``constant_csrf_token`` is now permitted for test configuration purposes.
  See https://github.com/Pylons/pyramid_beaker/pull/13 for more information.

0.7 (2012-07-10)
----------------

- Python 3.2 compatibility.

- Dropped Jython support (may be re-added after Jython releases a 2.7-
  compatible version).

- Dropped Python 2.5 support.

- Made compatible with zope.interface 4.0.0 (symptom: ``TypeError: Class
  advice impossible in Python3.  Use the @implementer class decorator
  instead.`` when run under Python 2 or *or* Python 3)

- Add ``dev`` and ``docs`` setup.py aliases.

0.6.1 (2011-11-19)
------------------

- Beaker 1.6+ compat: added support for ``key_length`` and ``enabled``
  options.

0.6 (2011-11-12)
----------------

- Add a tox.ini for Jenkins testing.

- Add includeme function to ease application setup.

0.5 (2011-03-30)
----------------

- The default value for the ``cookie_on_exception`` parameter to
  ``pyramid_beaker.BeakerSessionFactoryConfig`` is now ``True``.  This means
  that when view code causes an exception to be raised, and the session has
  been mutated, a cookie will be sent back in the response.  Previously its
  default value was ``False``.

0.4 (2011-01-08)
----------------

- Allow cache regions to inherit URL configuration from main cache options.

- Respect a ``cookie_on_exception`` configuration value passed to the
  ``BeakerSessionFactoryConfig`` constructor.  If this value is passed and
  it's not false, session cookies will be set even on responses resulting
  from an exception.

- Update to new relationship between ``new_csrf_token`` and
  ``get_csrf_token`` as described in Pyramid 1.0a9 interfaces.

0.3 (2010-12-29)
----------------

- Support ISession flash/csrf API added to Pyramid 1.0a8+.

0.2 (2010-11-27)
----------------

- Add cache_region support via
  ``pyramid_beaker.set_cache_regions_from_settings`` (and documented).

0.1
---

- Initial release.


pyramid_beaker/tests.py
import unittest

class TestPyramidBeakerSessionObject(unittest.TestCase):
    def _makeOne(self, request, **options):
        from pyramid_beaker import BeakerSessionFactoryConfig
        return BeakerSessionFactoryConfig(**options)(request)

    def test_instance_conforms(self):
        from zope.interface.verify import verifyObject
        from pyramid.interfaces import ISession
        request = DummyRequest()
        session = self._makeOne(request)
        verifyObject(ISession, session)

    def test_callback(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session['fred'] = 42
        session.save()
        self.assertEqual(session.accessed(), True)
        self.assertTrue(len(request.callbacks) > 0)
        response= DummyResponse()
        request.callbacks[0](request, response)
        self.assertTrue(response.headerlist)

    def test_new(self):
        request = DummyRequest()
        session = self._makeOne(request)
        self.assertTrue(session.new)

    def test___setitem__calls_save(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session['a'] = 1
        self.assertEqual(session.__dict__['_dirty'], True)

    def test___delitem__calls_save(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session['a'] = 1
        del session.__dict__['_dirty']
        del session['a']
        self.assertEqual(session.__dict__['_dirty'], True)

    def test_changed(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session.changed()
        self.assertEqual(session.__dict__['_dirty'], True)

    def test_clear(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session['a'] = 1
        session.clear()
        self.assertFalse('a' in session)
        self.assertEqual(session.__dict__['_dirty'], True)

    def test_update(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session.update({'a':1}, b=2)
        self.assertTrue('a' in session)
        self.assertTrue('b' in session)
        self.assertEqual(session.__dict__['_dirty'], True)

    def test_setdefault(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session.setdefault('a', 'b')
        self.assertTrue('a' in session)
        self.assertEqual(session.__dict__['_dirty'], True)

    def test_pop(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session['a'] = 1
        session.__dict__['_dirty'] = False
        result = session.pop('a')
        self.assertFalse('a' in session)
        self.assertEqual(result, 1)
        self.assertEqual(session.__dict__['_dirty'], True)

    def test_popitem(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session['a'] = 1
        session.__dict__['_dirty'] = False
        result = session.popitem()
        self.assertNotEqual(result, None)
        self.assertEqual(session.__dict__['_dirty'], True)

    def test_flash_default(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session.flash('msg1')
        session.flash('msg2')
        self.assertEqual(session['_f_'], ['msg1', 'msg2'])

    def test_flash_mixed(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session.flash('warn1', 'warn')
        session.flash('warn2', 'warn')
        session.flash('err1', 'error')
        session.flash('err2', 'error')
        self.assertEqual(session['_f_warn'], ['warn1', 'warn2'])

    def test_pop_flash_default_queue(self):
        request = DummyRequest()
        session = self._makeOne(request)
        queue = ['one', 'two']
        session['_f_'] = queue
        result = session.pop_flash()
        self.assertEqual(result, queue)
        self.assertEqual(session.get('_f_'), None)

    def test_pop_flash_nodefault_queue(self):
        request = DummyRequest()
        session = self._makeOne(request)
        queue = ['one', 'two']
        session['_f_error'] = queue
        result = session.pop_flash('error')
        self.assertEqual(result, queue)
        self.assertEqual(session.get('_f_error'), None)

    def test_peek_flash_default_queue(self):
        request = DummyRequest()
        session = self._makeOne(request)
        queue = ['one', 'two']
        session['_f_'] = queue
        result = session.peek_flash()
        self.assertEqual(result, queue)
        self.assertEqual(session.get('_f_'), queue)

    def test_peek_flash_nodefault_queue(self):
        request = DummyRequest()
        session = self._makeOne(request)
        queue = ['one', 'two']
        session['_f_error'] = queue
        result = session.peek_flash('error')
        self.assertEqual(result, queue)
        self.assertEqual(session.get('_f_error'), queue)

    def test_new_csrf_token(self):
        request = DummyRequest()
        session = self._makeOne(request)
        token = session.new_csrf_token()
        self.assertEqual(token, session['_csrft_'])
        # make sure its not a bytestring on py3
        self.assertTrue(str(token) == token)

    def test_get_csrf_token(self):
        request = DummyRequest()
        session = self._makeOne(request)
        session['_csrft_'] = 'token'
        token = session.get_csrf_token()
        self.assertEqual(token, 'token')
        self.assertTrue('_csrft_' in session)

    def test_get_csrf_token_new(self):
        request = DummyRequest()
        session = self._makeOne(request)
        token = session.get_csrf_token()
        self.assertTrue(token)
        self.assertEqual(session['_csrft_'], token)

    def test_get_constant_csrf_token(self):
        constant_token = 'FOO'
        request = DummyRequest()
        session = self._makeOne(request, constant_csrf_token=constant_token)
        token = session.get_csrf_token()
        self.assertEqual(token, constant_token)
        self.assertEqual(session['_csrft_'], token)



class Test_session_factory_from_settings(unittest.TestCase):
    def _callFUT(self, settings):
        from pyramid_beaker import session_factory_from_settings
        return session_factory_from_settings(settings)

    def test_it(self):
        settings = {'session.auto':'true', 'session.key':'foo'}
        factory = self._callFUT(settings)
        self.assertEqual(factory._options, {'auto':True, 'key':'foo'})

    def test_cookie_on_exception_true(self):
        settings = {'session.cookie_on_exception':'true'}
        factory = self._callFUT(settings)
        self.assertEqual(factory._cookie_on_exception, True)

    def test_cookie_on_exception_false(self):
        settings = {'session.cookie_on_exception':'false'}
        factory = self._callFUT(settings)
        self.assertEqual(factory._cookie_on_exception, False)

    def test_constant_csrf_token_set(self):
        settings = {'session.constant_csrf_token':'foo'}
        factory = self._callFUT(settings)
        self.assertEqual(factory._constant_csrf_token, 'foo')

    def test_constant_csrf_token_unset(self):
        settings = {}
        factory = self._callFUT(settings)
        self.assertEqual(factory._constant_csrf_token, False)


class DummyRequest:
    def __init__(self):
        self.callbacks = []
        self.environ = {}

    def add_response_callback(self, callback):
        self.callbacks.append(callback)

class DummyResponse:
    def __init__(self):
        self.headerlist = []

class Test_session_cookie_on_exception(unittest.TestCase):

    def _makeOne(self, request, **options):
        from pyramid_beaker import BeakerSessionFactoryConfig
        return BeakerSessionFactoryConfig(**options)(request)

    def test_default_cookie_on_exception_setting(self):
        request = DummyRequest()
        session = self._makeOne(request)
        self.assertEqual(session._cookie_on_exception, True)

    def test_cookie_on_exception_setting(self):
        request = DummyRequest()
        session = self._makeOne(request, cookie_on_exception=False)
        self.assertEqual(session._cookie_on_exception, False)

    def _assert_session_persisted(self, request, session, expected):
        # Checking response headers not likely best method of asserting
        # if Beaker's SessionObject.persist method was called.
        # Not sure of best method of doing this.
        request.exception = True
        session['use it'] = True
        response = DummyResponse()
        request.callbacks[0](request, response)
        self.assertEqual(len(response.headerlist) == 1, expected)

    def test_request_call_back_without_cookie_on_exception(self):
        request = DummyRequest()
        session = self._makeOne(request)
        self._assert_session_persisted(request, session, True)

    def test_request_call_back_with_cookie_on_exception(self):
        request = DummyRequest()
        session = self._makeOne(request, cookie_on_exception=False)
        self._assert_session_persisted(request, session, False)

class TestCacheConfiguration(unittest.TestCase):
    def _set_settings(self):
        return {'cache.regions':'default_term, second, short_term, long_term',
                'cache.type':'memory',
                'cache.second.expire':'1',
                'cache.short_term.expire':'60',
                'cache.default_term.expire':'300',
                'cache.long_term.expire':'3600',
                }

    def test_add_cache_no_regions(self):
        from pyramid_beaker import set_cache_regions_from_settings
        import beaker
        settings = self._set_settings()
        beaker.cache.cache_regions = {}
        settings['cache.regions'] = ''
        set_cache_regions_from_settings(settings)
        self.assertEqual(beaker.cache.cache_regions, {})

    def test_add_cache_single_region_no_expire(self):
        from pyramid_beaker import set_cache_regions_from_settings
        import beaker
        settings = self._set_settings()
        beaker.cache.cache_regions = {}
        settings['cache.regions'] = 'default_term'
        del settings['cache.default_term.expire']
        set_cache_regions_from_settings(settings)
        default_term = beaker.cache.cache_regions.get('default_term')
        self.assertEqual(default_term,
                         {'url': None, 'expire': 60, 'type': 'memory',
                          'lock_dir': None, 'data_dir': None, 'enabled': True,
                          'key_length': 250,})

    def test_add_cache_multiple_region(self):
        from pyramid_beaker import set_cache_regions_from_settings
        import beaker
        settings = self._set_settings()
        beaker.cache.cache_regions = {}
        settings['cache.regions'] = 'default_term, short_term'
        settings['cache.lock_dir'] = 'foo'
        settings['cache.short_term.expire'] = '60'
        settings['cache.default_term.type'] = 'file'
        settings['cache.default_term.expire'] = '300'
        settings['cache.default_term.enabled'] = 'false'
        set_cache_regions_from_settings(settings)
        default_term = beaker.cache.cache_regions.get('default_term')
        short_term = beaker.cache.cache_regions.get('short_term')
        self.assertEqual(short_term.get('expire'),
                         int(settings['cache.short_term.expire']))
        self.assertEqual(short_term.get('lock_dir'), settings['cache.lock_dir'])
        self.assertEqual(short_term.get('type'), 'memory')
        self.assertTrue(short_term.get('enabled'))

        self.assertEqual(default_term.get('expire'),
                         int(settings['cache.default_term.expire']))
        self.assertEqual(default_term.get('lock_dir'),
                         settings['cache.lock_dir'])
        self.assertEqual(default_term.get('type'),
                         settings['cache.default_term.type'])
        self.assertFalse(default_term.get('enabled'))

    def test_region_inherit_url(self):
        from pyramid_beaker import set_cache_regions_from_settings
        import beaker
        settings = self._set_settings()
        beaker.cache.cache_regions = {}
        settings['cache.regions'] = 'default_term, short_term'
        settings['cache.lock_dir'] = 'foo'
        settings['cache.url'] = '127.0.0.1'
        settings['cache.short_term.expire'] = '60'
        settings['cache.default_term.type'] = 'file'
        settings['cache.default_term.expire'] = '300'
        set_cache_regions_from_settings(settings)
        default_term = beaker.cache.cache_regions.get('default_term')
        short_term = beaker.cache.cache_regions.get('short_term')
        self.assertEqual(short_term.get('url'), settings['cache.url'])
        self.assertEqual(default_term.get('url'), settings['cache.url'])

    def test_region_inherit_enabled(self):
        from pyramid_beaker import set_cache_regions_from_settings
        import beaker
        settings = self._set_settings()
        settings['cache.enabled'] = 'false'
        beaker.cache.cache_regions = {}
        set_cache_regions_from_settings(settings)
        default_term = beaker.cache.cache_regions.get('default_term')
        short_term = beaker.cache.cache_regions.get('short_term')
        self.assertFalse(short_term.get('enabled'))
        self.assertFalse(default_term.get('enabled'))

class TestIncludeMe(unittest.TestCase):
    def test_includeme(self):
        from pyramid.interfaces import ISessionFactory
        from pyramid import testing
        from pyramid_beaker import includeme
        config = testing.setUp(settings={})
        includeme(config)
        session_factory = config.registry.queryUtility(ISessionFactory)
        self.assertEqual(str(session_factory),
                "<class 'pyramid_beaker.PyramidBeakerSessionObject'>")


pyramid_beaker/__init__.py
import os

from beaker import cache
from beaker.session import SessionObject
from beaker.util import coerce_cache_params
from beaker.util import coerce_session_params

from pyramid.interfaces import ISession
from pyramid.settings import asbool
from zope.interface import implementer

from binascii import hexlify


def BeakerSessionFactoryConfig(**options):
    """ Return a Pyramid session factory using Beaker session settings
    supplied directly as ``**options``"""

    class PyramidBeakerSessionObject(SessionObject):
        _options = options
        _cookie_on_exception = _options.pop('cookie_on_exception', True)
        _constant_csrf_token = _options.pop('constant_csrf_token', False)

        def __init__(self, request):
            SessionObject.__init__(self, request.environ, **self._options)
            def session_callback(request, response):
                exception = getattr(request, 'exception', None)
                if (
                    (exception is None or self._cookie_on_exception)
                    and self.accessed()
                ):
                    self.persist()
                    headers = self.__dict__['_headers']
                    if headers['set_cookie'] and headers['cookie_out']:
                        response.headerlist.append(
                            ('Set-Cookie', headers['cookie_out']))
            request.add_response_callback(session_callback)

        # ISession API

        @property
        def new(self):
            return self.last_accessed is None

        changed = SessionObject.save

        # modifying dictionary methods

        @call_save
        def clear(self):
            return self._session().clear()

        @call_save
        def update(self, d, **kw):
            return self._session().update(d, **kw)

        @call_save
        def setdefault(self, k, d=None):
            return self._session().setdefault(k, d)

        @call_save
        def pop(self, k, d=None):
            return self._session().pop(k, d)

        @call_save
        def popitem(self):
            return self._session().popitem()

        __setitem__ = call_save(SessionObject.__setitem__)
        __delitem__ = call_save(SessionObject.__delitem__)

        # Flash API methods
        def flash(self, msg, queue='', allow_duplicate=True):
            storage = self.setdefault('_f_' + queue, [])
            if allow_duplicate or (msg not in storage):
                storage.append(msg)

        def pop_flash(self, queue=''):
            storage = self.pop('_f_' + queue, [])
            return storage

        def peek_flash(self, queue=''):
            storage = self.get('_f_' + queue, [])
            return storage

        # CSRF API methods
        def new_csrf_token(self):
            token = (self._constant_csrf_token
                     or hexlify(os.urandom(20)).decode('ascii'))
            self['_csrft_'] = token
            return token

        def get_csrf_token(self):
            token = self.get('_csrft_', None)
            if token is None:
                token = self.new_csrf_token()
            return token

    return implementer(ISession)(PyramidBeakerSessionObject)


def call_save(wrapped):
    """ By default, in non-auto-mode beaker badly wants people to
    call save even though it should know something has changed when
    a mutating method is called.  This hack should be removed if
    Beaker ever starts to do this by default. """
    def save(session, *arg, **kw):
        value = wrapped(session, *arg, **kw)
        session.save()
        return value
    save.__doc__ = wrapped.__doc__
    return save


def session_factory_from_settings(settings):
    """ Return a Pyramid session factory using Beaker session settings
    supplied from a Paste configuration file"""
    prefixes = ('session.', 'beaker.session.')
    options = {}

    # Pull out any config args meant for beaker session. if there are any
    for k, v in settings.items():
        for prefix in prefixes:
            if k.startswith(prefix):
                option_name = k[len(prefix):]
                if option_name == 'cookie_on_exception':
                    v = asbool(v)
                options[option_name] = v

    options = coerce_session_params(options)
    return BeakerSessionFactoryConfig(**options)

def set_cache_regions_from_settings(settings):
    """ Add cache support to the Pylons application.

    The ``settings`` passed to the configurator are used to setup
    the cache options. Cache options in the settings should start
    with either 'beaker.cache.' or 'cache.'.

    """
    cache_settings = {'regions':None}
    for key in settings.keys():
        for prefix in ['beaker.cache.', 'cache.']:
            if key.startswith(prefix):
                name = key.split(prefix)[1].strip()
                cache_settings[name] = settings[key].strip()
    coerce_cache_params(cache_settings)

    if 'enabled' not in cache_settings:
        cache_settings['enabled'] = True

    regions = cache_settings['regions']
    if regions:
        for region in regions:
            if not region: continue
            region_settings = {
                'data_dir': cache_settings.get('data_dir'),
                'lock_dir': cache_settings.get('lock_dir'),
                'expire': cache_settings.get('expire', 60),
                'enabled': cache_settings['enabled'],
                'key_length': cache_settings.get('key_length', 250),
                'type': cache_settings.get('type'),
                'url': cache_settings.get('url'),
            }
            region_prefix = '%s.' % region
            region_len = len(region_prefix)
            for key in list(cache_settings.keys()):
                if key.startswith(region_prefix):
                    region_settings[key[region_len:]] = cache_settings.pop(key)
            coerce_cache_params(region_settings)
            cache.cache_regions[region] = region_settings

def includeme(config):
    session_factory = session_factory_from_settings(config.registry.settings)
    config.set_session_factory(session_factory)
    set_cache_regions_from_settings(config.registry.settings)
